### PR TITLE
Allow Forcfully to unlock PortLock on Operator

### PR DESCRIPTION
**Pull request by DmitryDodzin**

### CODE REVIEW

File: mirrord-schema.json [hunk 1]
```diff
@@ -591,6 +603,25 @@
         }
       ]
     },
+    "MultiSteal": {
+      "description": "(Operator Only): Allows overiding port locks\n\nCan be set to either `\"continue\"` (default) or `\"override\"`.\n\n- `\"continue\"`: Continue with normal execution - `\"override\"`: If port lock detected then override it with new lock and force close the original locking connection",
```
Comment (eyalb181) [inline mirrord-schema.json:607]:
```suggestion
      "description": "(Operator Only): Allows overriding port locks\n\nCan be set to either `\"continue\"` (default) or `\"override\"`.\n\n- `\"continue\"`: Continue with normal execution - `\"override\"`: If port lock detected then override it with new lock and force close the original locking connection.",
```

Comment (eyalb181) [reply mirrord-schema.json:607]: Also, default should be abort with an error message

File: mirrord/cli/src/config.rs [hunk 2]
```diff
@@ -183,6 +184,10 @@ pub(super) struct ExecArgs {
     /// Create a trace file of errors for debugging.
     #[arg(long)]
     pub capture_error_trace: bool,
+
+    /// Allows overiding port lock
```
Comment (eyalb181) [inline mirrord/cli/src/config.rs:188]:
```suggestion
    /// Allows overriding port lock
```

File: mirrord-schema.json [hunk 3]
```diff
@@ -526,6 +526,18 @@
             }
           ]
         },
+        "on_multi_steal": {
```
Comment (eyalb181) [inline mirrord-schema.json:529]:
```suggestion
        "on_concurrent_steal": {
```

File: mirrord/operator/src/client.rs [hunk 4]
```diff
@@ -85,7 +86,7 @@ impl OperatorApi {
     }
 
     /// Uses `MIRRORD_OPERATOR_SESSION_ID` to have a persistant session_id across child processes,
-    /// if env var is empty or missing random id will be generated and saved in env
+    /// if env var is empty or missing random id will be generated and saved in exnv
```
Comment (eyalb181) [inline mirrord/operator/src/client.rs:89]: Looks like a mistake

**Review by eyalb181 — CHANGES_REQUESTED** [review]
Few initial comments

**Review by Razz4780 — COMMENTED** [review]
LGTM apart from Eyal's suggestions

File: mirrord/cli/src/main.rs [hunk 5]
```diff
@@ -158,6 +158,13 @@ async fn exec(args: &ExecArgs, progress: &TaskProgress) -> Result<()> {
         std::env::set_var("MIRRORD_CAPTURE_ERROR_TRACE", "true");
     }
 
+    if let Some(on_multi_steal) = &args.on_multi_steal {
```
Comment (aviramha) [inline mirrord/cli/src/main.rs:161]: we should stop adding more envs..

**Comment by aviramha** [comment]
another thing that came to mind - we shouldn't do this if http filter is being used (i.e only "All" steal should cause this behavior)

File: mirrord-schema.json [hunk 6]
```diff
@@ -281,6 +281,32 @@
       },
       "additionalProperties": false
     },
+    "ConcurrentSteal": {
+      "description": "(Operator Only): Allows overiding port locks\n\nCan be set to either `\"continue\"` (default) or `\"override\"`.\n\n- `\"continue\"`: Continue with normal execution - `\"override\"`: If port lock detected then override it with new lock and force close the original locking connection",
```
Comment (eyalb181) [inline mirrord-schema.json:285]:
```suggestion
      "description": "(Operator Only): Allows overriding port locks.\n\nCan be set to either `\"continue\"` or `\"override\"`.\n\n- `\"continue\"`: Continue with normal execution - `\"override\"`: If port lock detected then override it with new lock and force close the original locking connection.",
```

File: mirrord-schema.json [hunk 7]
```diff
@@ -281,6 +281,32 @@
...
+        },
+        {
+          "description": "<!--${internal}--> ### continue\n\nContinue with normal execution",
+          "type": "string",
+          "enum": [
+            "continue"
+          ]
+        },
+        {
+          "description": "<!--${internal}--> ### abort\n\nAbort Execution",
```
Comment (eyalb181) [inline mirrord-schema.json:302]:
```suggestion
          "description": "<!--${internal}--> ### abort\n\nAbort Execution when trying to steal traffic from a target whose traffic is already being stolen.",
```

File: mirrord-schema.json [hunk 8]
```diff
@@ -598,6 +624,18 @@
             }
           ]
         },
+        "on_concurrent_steal": {
+          "title": "on_concurrent_steal",
+          "description": "(Operator Only): if value of override is passed then current later will override any PortLocks and force close other connections",
```
Comment (eyalb181) [inline mirrord-schema.json:629]: current later?

File: mirrord/cli/src/error.rs [hunk 9]
```diff
@@ -53,6 +53,27 @@ pub(crate) enum CliError {
     {GENERAL_HELP}"#
     ))]
     OperatorConnectionFailed(#[from] OperatorApiError),
+    #[error("Failed to connect to the operator. Someone else is performing network stealing on requested target")]
```
Comment (eyalb181) [inline mirrord/cli/src/error.rs:56]:
```suggestion
    #[error("Failed to connect to the operator. Someone else is stealing traffic from the requested target")]
```

File: mirrord/cli/src/error.rs [hunk 10]
```diff
@@ -53,6 +53,27 @@ pub(crate) enum CliError {
...
+    #[diagnostic(help(
+        r#"
+    There exists a port-lock on target, if you want to run anyway please set the following:
```
Comment (eyalb181) [inline mirrord/cli/src/error.rs:59]:
```suggestion
    If you want to run anyway please set the following:
```

File: mirrord/config/src/feature/network/incoming.rs [hunk 11]
```diff
@@ -95,6 +95,13 @@ impl MirrordConfig for IncomingFileConfig {
                     .transpose()?
                     .unwrap_or_default(),
                 http_header_filter: HttpHeaderFilterFileConfig::default().generate_config()?,
+                on_concurrent_steal: FromEnv::new("MIRRORD_OPERATOR_ON_MULTI_STEAL")
```
Comment (eyalb181) [inline mirrord/config/src/feature/network/incoming.rs:98]:
```suggestion
                on_concurrent_steal: FromEnv::new("MIRRORD_OPERATOR_ON_CONCURRENT_STEAL")
```

**Review by eyalb181 — CHANGES_REQUESTED** [review]

**Review by eyalb181 — APPROVED** [review]

**Pull request merged**